### PR TITLE
vm-cycle: use exponential backoff when pause due to IO error

### PR DESCRIPTION
**Pull request by fossedihelm**
**What this PR does / why we need it**:
Currently, once a VM is Paused due to IO error, the VM status is flipping between Paused and Running multiple times per second. This because the virt-handler tries to restart the VM, and if the IOError is not resolved, the latter will be paused again.
In order to limit the impact to the apiserver, when a domain IOError occurs, an increasingly exponential backoff will be applied before retrying to start the VM again.

Co-authored-by: fossedihelm <ffossemo@redhat.com>
Co-authored-by: lzang <zangli@google.com>

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes [https://bugzilla.redhat.com/show_bug.cgi?id=2029391](https://bugzilla.redhat.com/show_bug.cgi?id=2029391)

**Special notes for your reviewer**:
This is a new PR, started from https://github.com/kubevirt/kubevirt/pull/8562

**Release note**:

```release-note
Use an increasingly exponential backoff before retrying to start the VM, when an I/O error occurs.
```

### CODE REVIEW

**Comment by iholder-redhat** [comment]
Some questions / thoughts:
1. Let's say you `Queue.AddAfter(key, delay)`, say with a delay that is 1 minute. Now let's say that after 10 seconds the VM changes somehow hence the informer's event handlers wakes up and requeues the VM. I'm pretty sure that the VM would be requeued whenever it can, and won't be holded until 1 minutes has passed. Is that right?
2. Maybe instead of returning, you can make sure that `shouldUpdate` remains false and continue the execution. This might be important if some other processing needs to occur. Something even better (that might even be crucial) would be to allow updating stuff that aren't related to IO. I'm not sure where is the exact place in which we deal with io errors and set the new phase, but perhaps this specific place can skip its execution.
3. You basically implemented an exponential backoff mechanism. I wonder - isn't there a public tool, perhaps something provided by client-go or something similar from the kubernetes echosystem - that we can use? After all, this is very similar to what `AddRateLimited` does, right? WDYT?

**Comment by fossedihelm** [comment]
> Some questions / thoughts:
> 
> 1. Let's say you `Queue.AddAfter(key, delay)`, say with a delay that is 1 minute. Now let's say that after 10 seconds the VM changes somehow hence the informer's event handlers wakes up and requeues the VM. I'm pretty sure that the VM would be requeued whenever it can, and won't be holded until 1 minutes has passed. Is that right?

IIUC your scenario yes. For example: if the vmi is paused and the user `unpause` it manually, the vmi will not be holded due to the backoff.

> 2. Maybe instead of returning, you can make sure that `shouldUpdate` remains false and continue the execution. This might be important if some other processing needs to occur. Something even better (that might even be crucial) would be to allow updating stuff that aren't related to IO. I'm not sure where is the exact place in which we deal with io errors and set the new phase, but perhaps this specific place can skip its execution.

Not returning could be a good advice. Let me try it :)
About the IO error: it is qemu itself that change the domain status when the IO error occurs, we just react to the domain change, enqueueing the vmi to update it.

> 3. You basically implemented an exponential backoff mechanism. I wonder - isn't there a public tool, perhaps something provided by client-go or something similar from the kubernetes echosystem - that we can use? After all, this is very similar to what `AddRateLimited` does, right? WDYT?

Yes, there is https://pkg.go.dev/k8s.io/client-go/util/flowcontrol which was used in the first implementation https://github.com/kubevirt/kubevirt/pull/8562.
Basically the main reason why we implement a custom exponential backoff is that the `flowcontrol` doesn't obsolete the entry until twice the max backoff time, which means if a vm recovered in one retry, and then fail again within twice of the max backoff time, it will inherit the old backoff behavior.
While writing, it comes in my mind that I can use it with some custom adjustment. Is it something that can fit this request?

Thanks so much!

**Comment by alicefr** [comment]
This is just an idea. But why don't we wait before setting the VMI status to running? Basically, we unpause the VM in order to check if the IO error has been resolved. If it isn't the case, it returns to Paused. This is the reason why there is this annoying flipping between states. What I'm suggesting is to try to unpause the guest, wait a couple of seconds, and if the libvirt domain remains running, then we change the VMI status to Running. This might avoid this continuous changing state.
However, I still agree that increasing the time for check (unpausing the guest) makes a lot of sense.

**Comment by fossedihelm** [comment]
> This is just an idea. But why don't we wait before setting the VMI status to running? Basically, we unpause the VM in order to check if the IO error has been resolved. If it isn't the case, it returns to Paused. This is the reason why there is this annoying flipping between states. What I'm suggesting is to try to unpause the guest, wait a couple of seconds, and if the libvirt domain remains running, then we change the VMI status to Running. This might avoid this continuous changing state. However, I agree that increasing the time for check (unpausing the guest) makes a lot of sense.

@alicefr Thanks for your feedback! I think that the root problem is not that the vmi is flipping between running and paused, but the fact that this is done continuously which could impact on the apiserver. Postponing the update of the vmi could result in non-synchrony between domain and vmi, and, in general, a sort of information hidden. I mean, IMHO, I prefer to see the vmi status flipping, which gives me the feedback that we are trying to restart it, as long as this is not done multiple times in one second, but with a reasonable cadence. WDYT?

**Comment by alicefr** [comment]
> > This is just an idea. But why don't we wait before setting the VMI status to running? Basically, we unpause the VM in order to check if the IO error has been resolved. If it isn't the case, it returns to Paused. This is the reason why there is this annoying flipping between states. What I'm suggesting is to try to unpause the guest, wait a couple of seconds, and if the libvirt domain remains running, then we change the VMI status to Running. This might avoid this continuous changing state. However, I agree that increasing the time for check (unpausing the guest) makes a lot of sense.
> 
> @alicefr Thanks for your feedback! I think that the root problem is not that the vmi is flipping between running and paused, but the fact that this is done continuously which could impact on the apiserver. Postponing the update of the vmi could result in non-synchrony between domain and vmi, and, in general, a sort of information hidden. I mean, IMHO, I prefer to see the vmi status flipping, which gives me the feedback that we are trying to restart it, as long as this is not done multiple times in one second, but with a reasonable cadence. WDYT?

Yes, I make sense, thanks!

**Comment by fossedihelm** [comment]
@iholder-redhat I pushed the change which set `shouldUpdate` to false instead of returning. Thanks!

File: pkg/virt-handler/retry_manager_test.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("virt-handler retry manager", Ordered, func() {
+	initialWait, maxWait, maxFailResponseTime := 200*time.Millisecond, 5*time.Second, 1*time.Second
+	buffer := 10 * time.Millisecond
+	var retryManager *FailRetryManager
+	key := "c4ab4ae0-db63-45d8-aa0f-fc53dc84bdab"
```
Comment (iholder-redhat) [inline pkg/virt-handler/retry_manager_test.go:33]: Can you turn these to consts?
```suggestion
	const initialWait, maxWait, maxFailResponseTime = 200*time.Millisecond, 5*time.Second, 1*time.Second
	const buffer = 10 * time.Millisecond
	var retryManager *FailRetryManager
	const key = "c4ab4ae0-db63-45d8-aa0f-fc53dc84bdab"
```

File: pkg/virt-handler/retry_manager_test.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+	Context("retry manager", func() {
```
Comment (iholder-redhat) [inline pkg/virt-handler/retry_manager_test.go:35]: I think you can get rid of this `Context`, but if you decide to keep it, please either initialize retryManager inside or move the `BeforeEach()` to outside of the `Context`.

File: pkg/virt-handler/retry_manager_test.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+		BeforeEach(func() {
+			retryManager = NewFailRetryManager("test", initialWait, maxWait, maxFailResponseTime)
+		})
+
+		It("Should not delay if no failure occurred", func() {
+			shouldDelay, _ := retryManager.ShouldDelay(key, func() bool { return false })
```
Comment (iholder-redhat) [inline pkg/virt-handler/retry_manager_test.go:42]: nit: Perhaps you can have a helper function for these fake false/true functions:
```go
func getBoolFunc(b bool) func() bool {
	return func() bool {
		return b
	}
}
```

Then use it like that:
```go
retryManager.ShouldDelay(key, getBoolFunc(false))
```

Comment (iholder-redhat) [reply pkg/virt-handler/retry_manager_test.go:42]: Another alternative:
```go
shouldDelay := func(key string, isFailure bool) (bool, time.Duration) {
	return retryManager.ShouldDelay(key, func() bool { return isFailure })
}
```

Then:
```go
shouldDelay(key, false)
```

File: pkg/virt-handler/retry_manager_test.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+		})
+
+		It("Should do exponential backoff on failure retries", func() {
+			shouldDelay, _ := retryManager.ShouldDelay(key, func() bool { return true })
+			Expect(shouldDelay).To(BeFalse())
+			Expect(retryManager.retryStates[key].lastRunFailed).To(BeFalse())
+			// Call again as a failure response to the above call.
+			shouldDelay, _ = retryManager.ShouldDelay(key, func() bool { return true })
+			Expect(shouldDelay).To(BeTrue())
+			Expect(retryManager.retryStates[key].lastRunFailed).To(BeTrue())
```
Comment (iholder-redhat) [inline pkg/virt-handler/retry_manager_test.go:64]: This logic is tested separately, so I think we don't need it here
```suggestion
			shouldDelay, _ := retryManager.ShouldDelay(key, func() bool { return true })
			shouldDelay, _ = retryManager.ShouldDelay(key, func() bool { return true })
```

File: pkg/virt-handler/retry_manager_test.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+			Expect(shouldDelay).To(BeFalse())
+			_, lastEventFirstFail := retryManager.GetFirstFailureTime(key)
+			Expect(lastEventFirstFail).ToNot(BeEquivalentTo(firstEventFirstFail))
+		})
+
+		It("Should be able to remove a key", func() {
+			retryManager.Remove(key)
+			found, _ := retryManager.GetFirstFailureTime(key)
+			Expect(found).To(BeFalse())
+		})
```
Comment (iholder-redhat) [inline pkg/virt-handler/retry_manager_test.go:95]: I see that the `Remove()` method is used only here. Why do we need it if the keys are being garbaged-collected?

**Review by iholder-redhat — COMMENTED** [review]
Great stuff @fossedihelm! Thank you!
Left some suggestions and nits

File: pkg/virt-handler/vm.go [hunk 6]
```diff
@@ -1725,11 +1731,20 @@ func (d *VirtualMachineController) defaultExecute(key string,
...
 			shouldUpdate = true
 		}
+
+		if shouldDelay, _delay := d.ioErrorRetryManager.ShouldDelay(string(vmi.UID),
+			func() bool {
+				return shouldUpdate && domainExists && domain.Status.Status == api.Paused && domain.Status.Reason == api.ReasonPausedIOError
+			}); shouldDelay {
+			shouldUpdate = false
+			delay = _delay
+			log.Log.Object(vmi).Infof("Delay vm update for %f seconds", delay.Seconds())
```
[the remark this answers is not in the document]
Comment (fossedihelm) [reply pkg/virt-handler/vm.go]: I'm not sure I understand what you're referring to. Can you be more specific? :)

File: pkg/virt-handler/vm.go [hunk 7]
```diff
@@ -1792,16 +1807,16 @@ func (d *VirtualMachineController) defaultExecute(key string,
...
 		if err != nil {
 			log.Log.Object(vmi).Reason(err).Error("Updating the VirtualMachineInstance status failed.")
-			return err
+			return err, delay
 		}
 	}
 
 	if syncErr != nil {
-		return syncErr
+		return syncErr, 0
```
[the remark this answers is not in the document]
Comment (fossedihelm) [reply pkg/virt-handler/vm.go]: Leftover, but since I follow your suggestion regarding avoid returning the duration, this is not important anymore :)

File: pkg/virt-handler/vm.go [hunk 8]
```diff
@@ -1725,11 +1731,20 @@ func (d *VirtualMachineController) defaultExecute(key string,
...
 		if vmi.Status.Phase == phase {
 			shouldUpdate = true
 		}
+
+		if shouldDelay, _delay := d.ioErrorRetryManager.ShouldDelay(string(vmi.UID),
+			func() bool {
+				return shouldUpdate && domainExists && domain.Status.Status == api.Paused && domain.Status.Reason == api.ReasonPausedIOError
+			}); shouldDelay {
+			shouldUpdate = false
+			delay = _delay
```
[the remark this answers is not in the document]
Comment (fossedihelm) [reply pkg/virt-handler/vm.go]: Agree, this will prevent some changes! Done

File: pkg/virt-handler/retry_manager.go [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+	return &FailRetryManager{
+		name:                name,
+		maxFailResponseTime: maxFailResponseTime,
+		initialWait:         initialWait,
+		maxWait:             maxWait,
+		retryStates:         make(map[string]*retryState),
+	}
+}
+
+func (f *FailRetryManager) newState(now time.Time) *retryState {
```
[the remark this answers is not in the document]
Comment (fossedihelm) [reply pkg/virt-handler/retry_manager.go:69]: The `now` is parameterized in order to have always the same exact value. I know it is just fraction of milliseconds, but I tend to prefer it :)

File: pkg/virt-handler/retry_manager.go [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+}
+
+// Remove removes a key from manager.
+func (f *FailRetryManager) Remove(key string) {
+	f.stateLock.Lock()
+	defer f.stateLock.Unlock()
+
+	if _, ok := f.retryStates[key]; ok {
+		delete(f.retryStates, key)
+	}
```
[the remark this answers is not in the document]
Comment (fossedihelm) [reply pkg/virt-handler/retry_manager.go]: Thanks! Since the method was unused the method was deleted

File: pkg/virt-handler/retry_manager.go [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+	for {
+		select {
+		case <-ticker.C:
+			f.stateLock.Lock()
+			for key, state := range f.retryStates {
+				if !state.lastRunFailed && time.Now().After(state.lastRun.Add(f.maxFailResponseTime)) {
+					delete(f.retryStates, key)
+				}
+			}
+			f.stateLock.Unlock()
```
[the remark this answers is not in the document]
Comment (fossedihelm) [reply pkg/virt-handler/retry_manager.go]: Nice suggestion! Done!

File: pkg/virt-handler/retry_manager.go [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+
+	if !state.lastRunFailed {
+		if state.lastRun.Add(f.maxFailResponseTime).After(now) {
+			// This is a failure due to previous try.
+			state.lastRunFailed = true
+			return state.nextRun.After(now), state.nextRun.Sub(now)
+		} else {
+			// This is a new failure. Reset the status.
+			state = f.newState(now)
+			f.retryStates[key] = state
```
[the remark this answers is not in the document]
Comment (fossedihelm) [reply pkg/virt-handler/retry_manager.go]: Done

File: pkg/virt-handler/retry_manager_test.go [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+package virthandler
+
+import (
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("virt-handler retry manager", Ordered, func() {
```
[the remark this answers is not in the document]
Comment (fossedihelm) [reply pkg/virt-handler/retry_manager_test.go]: It is basically a leftover! Removed, thanks!

File: pkg/virt-handler/retry_manager_test.go (hunk 1 above)
Comment (fossedihelm) [reply pkg/virt-handler/retry_manager_test.go]: Done

File: pkg/virt-handler/retry_manager_test.go (hunk 2 above)
Comment (fossedihelm) [reply pkg/virt-handler/retry_manager_test.go]: Done

File: pkg/virt-handler/retry_manager_test.go (hunk 3 above)
Comment (fossedihelm) [reply pkg/virt-handler/retry_manager_test.go]: Thank you! I adopted the first alternative.

File: pkg/virt-handler/retry_manager_test.go (hunk 4 above)
Comment (fossedihelm) [reply pkg/virt-handler/retry_manager_test.go]: :+1:

File: pkg/virt-handler/retry_manager_test.go (hunk 5 above)
Comment (fossedihelm) [reply pkg/virt-handler/retry_manager_test.go]: Thanks! Initially I think I need it, but at the end it is unused. I removed it, and `GetFirstFailureTime` too (used only in test too)

**Comment by fossedihelm** [comment]
Thanks for your review @iholder-redhat! PTAL!

**Comment by iholder-redhat** [comment]
Awesome work!
/lgtm
/test pull-kubevirt-e2e-kind-1.22-sriov

**Review by iholder101 — APPROVED** [review]

File: pkg/virt-handler/retry_manager.go [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+		lastRun:       now,
+		nextRun:       now.Add(f.initialWait),
+		waitInterval:  f.initialWait,
+		lastRunFailed: false,
+	}
+}
+
+// ShouldDelay returns whether the retry on this failure should be delayed or not,
+// and if true return the duration of the delay as well.
+func (f *FailRetryManager) ShouldDelay(key string, isFailure func() bool) (bool, time.Duration) {
```
Comment (vladikr) [inline pkg/virt-handler/retry_manager.go:81]: As a general note, this is an area that will be interesting for anyone who will be trying to troubleshoot these failures. I think these state changes should be logged.

File: pkg/virt-handler/retry_manager.go [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+			state.lastRunFailed = true
+			return state.nextRun.After(now), state.nextRun.Sub(now)
+		} else {
+			// This is a new failure. Reset the status.
+			f.retryStates[key] = f.newState(now)
+			return false, 0
+		}
+	}
+
+	if !now.After(state.nextRun) {
```
Comment (vladikr) [inline pkg/virt-handler/retry_manager.go:109]: Please add a comment here to explain the logic. It should be easy to understand from a quick look at the code.

File: pkg/virt-handler/vm.go [hunk 16]
```diff
@@ -1760,6 +1764,15 @@ func (d *VirtualMachineController) defaultExecute(key string,
 		if vmi.Status.Phase == phase {
 			shouldUpdate = true
 		}
+
+		if shouldDelay, delay := d.ioErrorRetryManager.ShouldDelay(string(vmi.UID),
+			func() bool {
+				return shouldUpdate && domainExists && domain.Status.Status == api.Paused && domain.Status.Reason == api.ReasonPausedIOError
```
Comment (vladikr) [inline pkg/virt-handler/vm.go:1770]: Could you please separate this into its own function?

File: pkg/virt-handler/retry_manager.go (hunk 14 above)
Comment (fossedihelm) [reply pkg/virt-handler/retry_manager.go:82]: Thanks! Agreed, done!

File: pkg/virt-handler/retry_manager.go (hunk 15 above)
Comment (fossedihelm) [reply pkg/virt-handler/retry_manager.go:127]: You're right! I added some comments. PTAL

File: pkg/virt-handler/vm.go (hunk 16 above)
Comment (fossedihelm) [reply pkg/virt-handler/vm.go]: Done

**Comment by vladikr** [comment]
/approve
Thanks @fossedihelm
